### PR TITLE
ESQL: Add Warning tests for all remaining functions

### PR DESCRIPTION
**Pull request by bpintea**
This adds unit tests cases for all the functions that were missing tests checking on the correct generation of the Warning headers in case the execution raised an Exception that lead to a `null` result.

It extends the test cases with the verification of the exceptions occurring on folding. This is required since some evaluators are constructed based on the function input (literal or expression) and the processing exception can occur at this point already (so not in the evaluator catching the exception and turning it into a warning).

Related: #102488

### CODE REVIEW

**Review by costin — APPROVED** [review]
LGTM.
I'd still wait for a review from @nik9000 or @not-napoleon .

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java
Comment (not-napoleon) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java:474]: Maybe I'm just confused, but I feel like this check is contradictory to the comment above it.  Does the comment need to be rewritten? or just added to?

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java
Comment (not-napoleon) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java:1192]: Huh.  I'm not sure why we need this; or rather, I'm not sure why we didn't need this sooner if we need it at all. When are the folding warnings different from the evaluator warnings? is it possible we can merge down to a single warnings/exceptions function some how?

**Review by not-napoleon — COMMENTED** [review]
As discussed in slack, I think a key outcome for this work is wiring up `anyNullIsNull` and `errorsForCasesWithoutExamples` to all these function tests.  I'd also like to move more of these cases to the data generation functions in `TestCaseSupplier`; that will help us cover more edge cases long term (e.g., I don't want to have to remember that everything which accepts a double needs to check negative zero, I just want that to happen automatically)

I'm indifferent as to if we merge this PR and do more follow up work, or if we address these concerns in this PR.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java
Comment (bpintea) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/TestCaseSupplier.java:1170]: > I'm not sure why we need this; or rather, I'm not sure why we didn't need this sooner if we need it at all.

It's been needed, IMO, we just didn't do it (for instance, have a look at 0-avoiding random number generation in `DivTests`, but no 0 devisor test.. which this PR also adds).
I've added this now because some functions (like `DateParseTests.java`) generate the same-ish message both in `Warning`s (when run through the execution engine) and folding failures (when the expression can be executed / folded before execution) and the folding tests cannot (and should not) be disabled.

**Comment by bpintea** [comment]
> I'm indifferent as to if we merge this PR and do more follow up work, or if we address these concerns in this PR.

@not-napoleon, after further considering it, I think it'd make sense to merge this PR as is, as it only extends existing suites. 
Updating these - and other suites not touched by this PR - to wire the cases through `errorsForCasesWithoutExamples(anyNullIsNull(...))` makes a lot of sense, but would extend the PR beyond its initial scope (i.e. checking all functions for correct `Warning`s).

**Pull request merged**
